ocaml/typing/typecore.ml

```diff
@@ -7674,6 +7784,57 @@ and type_tuple ~loc ~env ~(expected_mode : expected_mode) ~ty_expected
     exp_attributes = attributes;
     exp_env = env }
 
+and type_unboxed_tuple ~loc ~env ~(expected_mode : expected_mode) ~ty_expected
+      ~explanation ~attributes sexpl =
+  Jane_syntax_parsing.assert_extension_enabled ~loc Layouts
+    Language_extension.Beta;
+  let arity = List.length sexpl in
+  assert (arity >= 2);
+  (* elements must be representable *)
+  let labels_types_and_sorts =
+    List.map (fun (label, _) ->
+      let jkind, sort = Jkind.of_new_sort_var ~why:Unboxed_tuple_element in
+      label, newgenvar jkind, sort)
+    sexpl
+  in
+  let labeled_subtypes =
+    List.map (fun (l, t, _) -> (l, t)) labels_types_and_sorts
+  in
+  let to_unify = newgenty (Tunboxed_tuple labeled_subtypes) in
+  with_explanation explanation (fun () ->
+    unify_exp_types loc env to_unify (generic_instance ty_expected));
+
+  let argument_modes =
+    (* XXX modes *)
+    match expected_mode.tuple_modes with
+    (* CR zqian: improve the modes of opened labeled tuple pattern. *)
+    | Some tuple_modes when List.compare_length_with tuple_modes arity = 0 ->
+        tuple_modes
+    | _ ->
+        List.init arity (fun _ -> expected_mode.mode)
```

Comment:
Using `expected_mode.mode` while ignoring `tuple_modes` is dangerous. (See the comments at the definition of `type expected_mode`). We should use `as_single_mode expected_mode` here.

Speaking of which, could you copy `tuple_modes.ml` to `unboxed_tuple_modes.ml` and adapt the tests to unboxed tuples? Duplicative type checking logic deserves duplicative tests.

Reply:
Also as you noted, we should unify the type checking of regular tuple and unboxed tuple, as well as `Texp_tuple` and `Texp_unboxed_tuple`.

Reply:
Zesen and I discussed this out of band.  I adopted his suggestion to duplicate the tests (though I put them all in the same file to make it easier to compare the boxed and unboxed versions).  This turned up some mode bugs that he fixed, and also some missing cases in `parmatch.ml` which are fixed in [this commit](https://github.com/ocaml-flambda/flambda-backend/commit/0d85b523217c85b66ef8f2df40b9265fe4ee4dd1) (which probably should get a quick glance from someone besides me).

Reply:
Regarding unifying the typing of `Texp_tuple` and `Texp_unboxed_tuple` - this could be done, but I think it would make the code less clear.  They use different constructors everywhere and these constructors have different types (`Texp_tuple` takes a list of pairs while `Texp_unboxed_tuple` takes a list of triples) and need various special cases (e.g., the regular tuple case restricts everything to value, while the unboxed tuple case doesn't).  I tried but found it more confusing then helpful. In the future there will be fewer special cases and we'll probably need a list of triples in both cases, so I propose to wait until then.